packages/shared/lib/services/connection.service.ts

```diff
@@ -2044,6 +2102,54 @@ class ConnectionService {
         }
     }
 
+    public getSignatureBasedCredentials(provider: ProviderSignatureBased, username: string, password: string): ServiceResponse<SignatureBasedCredentials> {
+        try {
+            let token: string;
+
+            if (provider.signature_protocol === 'WSSE') {
+                token = this.generateWsseCredentials(username, password);
+            } else {
+                throw new NangoError('unsupported_signature_protocol', { message: 'Signature protocol not supported' });
+            }
+
+            const expiresAt = new Date(Date.now() + provider.token.expires_in_ms);
+
+            const credentials: SignatureBasedCredentials = {
+                type: 'SIGNATURE_BASED',
+                username,
+                password,
+                token,
+                expires_at: expiresAt
+            };
+
+            return { success: true, error: null, response: credentials };
+        } catch (err) {
+            const error = new NangoError('signature_based_token_generation_error', { message: err instanceof Error ? err.message : 'unknown error' });
+            return { success: false, error, response: null };
+        }
+    }
+
+    private generateWsseCredentials(username: string, password: string): string {
+        try {
+            const nonce = crypto.randomBytes(16).toString('hex');
+
+            const timestamp = new Date().toISOString();
+
+            const sha1Hash = crypto
+                .createHash('sha1')
+                .update(nonce + timestamp + password)
+                .digest('hex');
+
+            const passwordDigest = Buffer.from(sha1Hash, 'hex').toString('base64');
+
+            const token = `UsernameToken Username="${username}", PasswordDigest="${passwordDigest}", Nonce="${nonce}", Created="${timestamp}"`;
+
+            return token;
+        } catch (err) {
+            throw new NangoError('wsse_token_generation_error', { message: err instanceof Error ? err.message : 'unknown error' });
+        }
+    }
```

Comment:
I would have this function in its own file in `lib/signature/wsse` for instance. It would be easier to find and once we add a couple of other signature methods, the current file might become a bit messy

Reply:
I have updated this with the latest [commit](https://github.com/NangoHQ/nango/pull/2942/commits/b4d3dd0ffcd24f1188b251597107f09a4e90eb13)